src/gpuarray_reduction.c

```diff
@@ -256,113 +801,492 @@ static int   maxandargmaxCheckargs              (maxandargmax_ctx*  ctx){
+			ctx->axisList[f++] = i;
+		}
+	}
+	memcpy(&ctx->axisList[f], ctx->reduxList, ctx->ndr * sizeof(*ctx->reduxList));
+	
+	
+	return reduxGenSource(ctx);
+}
+
+/**
+ * @brief Generate the kernel code for the reduction.
+ *
+ * @return GA_MEMORY_ERROR if not enough memory left; GA_NO_ERROR otherwise.
+ */
+
+static int   reduxGenSource                (redux_ctx*  ctx){
+	reduxAppendSource(ctx);
 	ctx->sourceCode = strb_cstr(&ctx->s);
 	if(!ctx->sourceCode){
-		return ctx->ret=GA_MEMORY_ERROR;
+		return reduxCleanup(ctx, GA_MEMORY_ERROR);
 	}
-
-	/* Return it. */
-	return ctx->ret=GA_NO_ERROR;
+	
+	return reduxIsLargeCodeModel(ctx) ? reduxCompileLarge(ctx):
+	                                    reduxCompileSmall(ctx);
 }
-static void  maxandargmaxAppendKernel           (maxandargmax_ctx*  ctx){
-	maxandargmaxAppendTypedefs         (ctx);
-	maxandargmaxAppendPrototype        (ctx);
-	strb_appends           (&ctx->s, "{\n");
-	maxandargmaxAppendOffsets          (ctx);
-	maxandargmaxAppendIndexDeclarations(ctx);
-	maxandargmaxAppendRangeCalculations(ctx);
-	maxandargmaxAppendLoops            (ctx);
-	strb_appends           (&ctx->s, "}\n");
+static void  reduxAppendSource             (redux_ctx*  ctx){
+	reduxAppendIncludes         (ctx);
+	reduxAppendTypedefs         (ctx);
+	reduxAppendFuncGetInitVal   (ctx);
+	reduxAppendFuncLoadVal      (ctx);
+	reduxAppendFuncReduxVal     (ctx);
+	reduxAppendFuncPreKernel    (ctx);
+	reduxAppendFuncKernel       (ctx);
+	reduxAppendFuncPostKernel   (ctx);
 }
-static void  maxandargmaxAppendTypedefs         (maxandargmax_ctx*  ctx){
+static void  reduxAppendIncludes           (redux_ctx*  ctx){
+	strb_appends(&ctx->s, "/* Includes */\n");
+	strb_appends(&ctx->s, "\n");
+	strb_appends(&ctx->s, "\n");
+	strb_appends(&ctx->s, "\n");
+}
+static void  reduxAppendTypedefs           (redux_ctx*  ctx){
 	strb_appends(&ctx->s, "/* Typedefs */\n");
-	strb_appendf(&ctx->s, "typedef %s     T;/* The type of the array being processed. */\n", ctx->dstMaxType);
-	strb_appendf(&ctx->s, "typedef %s     X;/* Index type: signed 32/64-bit. */\n",          ctx->dstArgmaxType);
+	strb_appendf(&ctx->s, "typedef %s     S;/* The type of the source array. */\n",                ctx->srcTypeStr);
+	strb_appendf(&ctx->s, "typedef %s     T;/* The type of the destination array. */\n",           ctx->dstTypeStr);
+	strb_appendf(&ctx->s, "typedef %s     A;/* The type of the destination argument array. */\n",  ctx->dstArgTypeStr);
+	strb_appendf(&ctx->s, "typedef %s     X;/* The type of the indices: signed 32/64-bit. */\n",   ctx->idxTypeStr);
+	strb_appendf(&ctx->s, "typedef %s     K;/* The type of the accumulator variable. */\n",        ctx->accTypeStr);
+	strb_appends(&ctx->s, "\n");
+	strb_appends(&ctx->s, "\n");
+	strb_appends(&ctx->s, "\n");
+}
+static void  reduxAppendFuncGetInitVal     (redux_ctx*  ctx){
+	strb_appends(&ctx->s, "/**\n");
+	strb_appends(&ctx->s, " * Initial value function.\n");
+	strb_appends(&ctx->s, " */\n");
```

Comment:
I don't know it outputting comments in a generated kernel is useful at all.  Perhaps you can keep this as an actual comment right here.